pkg/daemon/ceph/osd/kms/kms.go

```diff
@@ -114,6 +117,33 @@ func (c *Config) PutSecret(secretName, secretValue string) error {
 			return errors.Wrap(err, "failed to put secret in ibm key protect")
 		}
 	}
+	if c.IsKMIP() {
+		_, err := c.getKubernetesSecret(secretName)
+		if err == nil {
+			// if error is nil, secret exists, just return nil.
+			return nil
+		}
+		// if error is not found, continue with creation.
+		if !kerrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to check secret exists for %q", secretName)
+		}
+
+		kmip, err := InitKMIP(c.clusterSpec.Security.KeyManagementService.ConnectionDetails)
+		if err != nil {
+			return errors.Wrap(err, "failed to init kmip")
+		}
+
+		// register the key with kmip server.
+		uniqueIdentifier, err := kmip.registerKey(secretName, secretValue)
+		if err != nil {
+			return errors.Wrap(err, "failed to register secret in kmip")
+		}
+		// store the uniqueIdentifier in Kubernetes Secret.
+		err = c.storeSecretInKubernetes(secretName, uniqueIdentifier)
```

Comment:
> If we store the secret in kubernetes, we have lost the benefit of the KMS. 

We are storing the `uniqueIdentifier` associated with the secret key value.
The encryption key itself will be stored in kmip server.

>We only want to keep the secrets in the KMS, and query the KMS whenever a secret is needed. Why do we need to store the key in K8s? The secret could be read by anyone with access to the rook namespace.

This is because kmip only support key operations where `uniqueIdentifier` will be available post creation/registration and this cannot be set by the client.
We are storing this `uniqueIdentifier` in k8s secret.

All other kms currently supported still store their credentials in k8s secret, which could still be read by anyone with access to rook ns and they could connect to kms themselves to query the key.

Reply:
When retrieving the secrets from the kms like in `InitVault()` and `InitKeyProtect()` I don't see where a k8s secret is accessed, but agreed that if they have access to the rook namespace they could get access to the kms anyway since the cluster CR has the connection details for the kms.